### PR TITLE
Automatically generate handlers from Messengers

### LINKED ISSUE #2188: Check constructor argument types in Messengers and handlers
In [this topic on the forum](https://forum.pyro.ai/t/poutine-mask-intended-usage/1428/2), a user passed a mask tensor to `poutine.mask` as an argument `poutine.mask(mask)` rather than a keyword argument `poutine.mask(mask=mask)`, attempted to use the resulting `Messenger` as a context manager, and got an obscure error message as a result:
```
ValueError: Expected mask to be a boolean but got <class ‘NoneType’>
```
It would be good to raise more informative error messages in such cases that would make the issue clear (in this case, raising a `TypeError` or `ValueError` when initializing the `Messenger`'s `fn` attribute to something clearly not callable).

### PR DESCRIPTION
**Pull request by eb8680**
Resolves #2188.

This almost-pure refactoring PR:
1. Removes all the hand-defined boilerplate convenience functions like `poutine.trace` defined in `pyro/poutine/handlers.py` and moves or merges their docstrings into the corresponding `Messenger` classes.
2. Programatically generates new versions of these functions that behave identically (necessary to address #2188)
3. Adds a one-line check in the automatically generated versions that addresses #2188 
4. Adds a regression test that would have caught #2188 to `test/poutine/test_poutine.py`
5. Changes the name of `EnumerateMessenger` to `EnumMessenger` to avoid a name collision with the `enumerate` builtin during programmatic handler generation and be consistent with its preexisting alias `poutine.enum`

Almost all of the changes in this PR are copy-pasted docstrings and deleted boilerplate.

Tested:
- Pure refactoring exercised by existing tests
- Regression test in `test_poutine.py`
- Docs render correctly on my machine

### CODE REVIEW

File: pyro/poutine/handlers.py [hunk 1]
```diff
@@ -69,359 +70,76 @@
...
+                "{} is not callable, did you mean to pass it as a keyword arg?".format(fn))
+        msngr = msngr_cls(*args, **kwargs)
+        return msngr(fn) if fn is not None else msngr
+
+    return handler
+
+
+_re1 = re.compile('(.)([A-Z][a-z]+)')
+_re2 = re.compile('([a-z0-9])([A-Z])')
+for _msngr_cls in _msngrs:
```
Comment (eb8680) [inline pyro/poutine/handlers.py]: Here is the programmatic handler generation; compare to `pyro/distributions/torch.py`

File: pyro/poutine/handlers.py [hunk 2]
```diff
@@ -69,359 +70,76 @@
...
+    UnconditionMessenger,
+]
+
+
+def _make_handler(msngr_cls):
+
+    def handler(fn=None, *args, **kwargs):
+        if fn is not None and not callable(fn):
+            raise ValueError(
+                "{} is not callable, did you mean to pass it as a keyword arg?".format(fn))
```
Comment (eb8680) [inline pyro/poutine/handlers.py]: Here is the fix for #2188

File: pyro/poutine/handlers.py [hunk 3]
```diff
@@ -69,359 +70,76 @@
...
+
+    return handler
+
+
+_re1 = re.compile('(.)([A-Z][a-z]+)')
+_re2 = re.compile('([a-z0-9])([A-Z])')
+for _msngr_cls in _msngrs:
+    _handler_name = _re2.sub(
+        r'\1_\2', _re1.sub(r'\1_\2', _msngr_cls.__name__.split("Messenger")[0])).lower()
+    _handler = _make_handler(_msngr_cls)
```
Comment (fritzo) [inline pyro/poutine/handlers.py]: `_handler.__name__ = _handler_name`?

File: pyro/poutine/infer_config_messenger.py [hunk 4]
```diff
@@ -3,7 +3,13 @@
 
 class InferConfigMessenger(Messenger):
     """
-    Modifies contents of the infer kwarg at sample sites
+    Given a callable that contains Pyro primitive calls
```
Comment (fritzo) [inline pyro/poutine/infer_config_messenger.py]:
```
Given a callable `fn` that contains Pyro primitive calls
and a callable `config_fn` taking a trace ...
```

File: pyro/poutine/handlers.py (hunk 1 above)
Comment (eb8680) [reply pyro/poutine/handlers.py]: @neerajprad will this play nicely with serialization?  Do I need to use `functools.partial` here instead of a closure?

**Comment by eb8680** [comment]
> I noticed EnumerateMessenger/EnumMessenger is not documented (otherwise we would be introducing a breaking change). Should it be documented?

Probably, but this PR is just refactoring - all of the documentation changes here are from copy-pasting handler docstrings into messengers, not rewriting them.

File: pyro/poutine/handlers.py (hunk 1 above)
Comment (neerajprad) [reply pyro/poutine/handlers.py]: This should work fine! These will be available from the module namespace, so pickle shouldn't have a problem.

**Comment by eb8680** [comment]
I'll update the Sphinx config/doc generation for `pyro.poutine` in a followup PR to make sure all `Messenger`s are included.

File: pyro/poutine/block_messenger.py [hunk 5]
```diff
@@ -77,47 +74,38 @@ class BlockMessenger(Messenger):
...
     :param expose_fn: function that takes a site and returns True to expose the site
         or False/None to hide it.  If specified, all other parameters are ignored.
         Only specify one of hide_fn or expose_fn, not both.
-    :param bool hide_all: hide all sites
-    :param bool expose_all: expose all sites normally
-    :param list hide: list of site names to hide, rest will be exposed normally
-    :param list expose: list of site names to expose, rest will be hidden
-    :param list hide_types: list of site types to hide, rest will be exposed normally
-    :param list expose_types: list of site types to expose normally, rest will be hidden
+    :param hide: list of site names to hide
```
Comment (neerajprad) [inline pyro/poutine/block_messenger.py]: I think the previous docstring was more complete - `hide_all` and `expose_all` parameters are missing. Also maybe we should put in the type info back - some IDEs will do type checks based on the docstring.

Comment (neerajprad) [reply pyro/poutine/block_messenger.py]: Feel free to address this in your followup PR, btw.

Comment (eb8680) [reply pyro/poutine/block_messenger.py]: Good catch, I'll include this in the followup PR

**Pull request merged**